deeppavlov/pipeline_manager/pipegen.py

```diff
@@ -0,0 +1,273 @@
+
+    def pipeline_enumeration(self) -> Iterator:
+        """
+        Creates a primary set of pipelines using a self.main_config attribute.
+
+        Returns:
+            iterator of primary set of pipelines
+        """
+        for components in self.structure:
+            self.pipes.append(components)
+
+        return product(*self.pipes)
+
+    def _universial(self, pipe: tuple, ind: int):
+        new_config = dict(dataset_reader=deepcopy(self.dataset_reader),
+                          dataset_iterator=self.main_config['dataset_iterator'],
+                          chainer=self.chainer, train=self.train_config)
+        if 'metadata' in self.main_config.keys():
+            new_config['metadata'] = self.main_config['metadata']
+
+        chainer_components = list(pipe)
+        dataset_name = self.dataset_reader['data_path'].split('/')[-1]
+        chainer_components = self.change_load_path(chainer_components, ind, self.save_path, dataset_name,
+                                                   self.test_mode)
+        new_config['chainer']['pipe'] = chainer_components
+        ind += 1
+        return new_config
+
+    def pipeline_gen(self) -> Generator:
+        """
+        Creates a configs with a different set of hyperparameters based on the primary set of pipelines.
+
+        Returns:
+            iterator of final sets of configs (dicts)
+        """
+        p = 0
+        for i, pipe_var in enumerate(self.enumerator):
+            if self.mode == 'random':
+                for pipe_ in self.random_conf_gen(pipe_var):
+                    yield self._universial(pipe_, p)
+            else:
+                for pipe_ in self.grid_conf_gen(pipe_var):
+                    yield self._universial(pipe_, p)
+
+    # random generation
+    def random_conf_gen(self, pipe_components: List[dict]) -> Generator:
+        """
+        Creates a set of configs with a different set of hyperparameters using "random search".
+
+        Args:
+            pipe_components: list of dicts; config if components
+
+        Returns:
+            random search iterator
+        """
+        sample_gen = ParamsSearch(prefix="random", seed=42)
+        new_pipes = []
+        for component in pipe_components:
+            new_components = []
+            if component:
+                search = False
+                for key, item in component.items():
+                    if isinstance(item, dict):
+                        for item_key in item.keys():
+                            if item_key.startswith('random_'):
+                                search = True
+                                break
+
+                if search:
+                    for i in range(self.N):
+                        new_components.append(sample_gen.sample_params(**component))
+                    new_pipes.append(new_components)
+                else:
+                    new_components.append(component)
+                    new_pipes.append(new_components)
```

Comment:
if `new_pipes.append(new_components)` happens in both cases, it should not be inside an `if-else`

Reply:
Yes, of course.